Increase commbreakautowait and commbreakautowind max seconds to 60

DerickLebron opened the pull request:
Increase commbreakautowait and commbreakautowind max seconds from 10 to 60.

## Description
Allow max commbreakautowait time to be up to 60 seconds. 
Modify Edl.cpp autowait and autowind logic to handle the situation where the wait and or wind time results in the start time being greater than the end time.

## Motivation and Context
The mythtv backend usually flags the beginning of commercial breaks 20-30 seconds before the break starts. Kodi only allows a 10 second wait.

I'm not sure if this is an open issue, but I did find a user complaining about it in the link below when searching the web for the issue.
http://forum.kodi.tv/showthread.php?tid=257520
This user had the exact behavior wrong, but I believe was suffering from the same issue as me. I was not able to determine the exact behavior until looking at the code.

## How Has This Been Tested?
I am not able to test as I don't have a test environment.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
(The wiki does not specify a maximum value for these settings, but I would have found it useful if it did)

- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

MartijnKaijser commented:
@ksooo or @metaron-uk perhaps any comments?

metaron-uk commented on xbmc/cores/VideoPlayer/Edl.cpp:
```diff
@@ -685,9 +685,19 @@ bool CEdl::AddCut(Cut& cut)
+      int cutLength = cut.end - cut.start;
+      /* subtract the lesser of the cut length or the autowind from the end */
+      cut.end -= autowind > cutLength ? cutLength : autowind;
+    }
```
What about if (cut.start + autowait) < (cut.end), but (cut.start + autowait) > (cut.end - autowind)? (I've seen some very short commercial breaks late at night when the channel can't find any advertizers). Won't this result in a negative cut?

DerickLebron replied:
I'm hesitant to comment because of the don't comment rule, but I believe this is meaningful.
Unless my logic is flawed, this code would not result in a negative cut because the cutLength is recalculated as the remaining cutLength after the autowait is applied to the cut.start.
In the scenario outlined, cut.start would end up equal to cut.end, resulting in a 0 length cut.
The previous code would have handled this scenario by ignoring the autowind. 
I feel the better solution is to disregard the cut if the cutLengh < (autowait + autowind).

The scenario in which autowait > (cut.end - cut.start) is one that I also though would become more likely if the change to increase the maximum autowait and autowind values was accepted. I don't believe that scenario was handled by the original code.

Thanks for taking the time to look at this change even though I was not able to complete the checklist.

FernetMenta replied:
> I'm hesitant to comment because of the don't comment rule

@DerickLebron what do you mean? this is your PR, you can comment as much as you want

metaron-uk replied:
@DerickLebron sorry - you're right. Review in haste repent at leisure. :-(
By the way, if by any chance you're managing to use mythtv commercial detection successfully in the UK, can you post to the forum how you got it working and PM me a link? I'd be very interested.... :-).

DerickLebron replied:
@metaron-uk Sorry, I'm in the US.

metaron-uk reviewed (commented):
I don't use commskip myself although I do use a mythtv backend (in the UK I've never been able to get my mythtv backend to flag commercials reliably).

MartijnKaijser commented:
@ksooo what's your opinion in this?

ksooo commented:
> @ksooo what's your opinion in this?

 I've never used edls/commskip.

DerickLebron committed:
Max 60s commbreakautowait and commbreakautowind

Some users need to set my commbreakautowait to a value above 10 (30 in my case).
Here is a link to a thread I found while searching for why commbreakautowait was not working.
From looking at the code, then testing, I can see that it does work, but is limited to 10 seconds.
http://forum.kodi.tv/showthread.php?tid=257520

Update autowind and autowait logic

Update the autowait and autowind logic to handle the extended times.
This change will in effect cause a cut to be skipped when the autowait + autowind values are greater than the length of the cut, but setting the start and end of the cut to the same time.
The code will not allow the end time for the cut to be less than the start time.

Update xbmc/settings/AdvancedSettings.cpp

Allow negative comm skip time for increased flexibility to support intricacies of comm flagging with various backends.

Co-Authored-By: Andrew Fowler <metaron-uk@users.noreply.github.com>

Update xbmc/settings/AdvancedSettings.cpp

Allow negative comm skip time to allow for backends that are late at flagging commercial breaks.

Co-Authored-By: Andrew Fowler <metaron-uk@users.noreply.github.com>

phunkyfish commented:
@DerickLebron are you still interested in getting this PR merged?

DerickLebron commented:
@phunkyfish It would be nice if it could be. I understand if it can’t though because I wasn’t able to do proper testing. Thanks.

phunkyfish commented:
We could create a test build if you have some users who can test this?

Maybe search and forum and ping a few who have posted around edl/comskip.

DerickLebron commented:
@phunkyfish I think I can test if I have a build. I just don’t have an environment set up to build/debug.

phunkyfish commented:
Ok, what platform do you need?

metaron-uk commented:
happy to test (libreelec rpi2 build preferred, pvr.mythtv backend) or given a week or two can roll my own...

> On 8 Mar 2020, at 18:20, phunkyfish <notifications@github.com> wrote:
> 
> ﻿
> We could create a test build if you have some users who can test this?
> 
> Maybe search and forum and ping a few who have posted around edl/comskip.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

DerickLebron commented:
@phunkyfish windows would be easiest for me.

phunkyfish commented:
Cool, will post windows build shortly.

@MilhouseVH could you add this PR to the Rpi testbuilds so @metaron-uk can test also?

phunkyfish commented:
@DerickLebron win64 testbuild: https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20180925-63d4c7a5-patch-1-x64.exe

DerickLebron commented:
@phunkyfish thanks for getting this out there. The email this slipped past me the other day. I'm trying to test now, but I'm not seeing the mythtv pvr client or any PVR add-ons to add. The PVR client section isn't in the add-on categories. The repository looks the same as my windows app store version and libreelec. Sorry for the hassle, but I'm not sure how to get it set up.

phunkyfish commented:
Are you not able to install from repository?

DerickLebron commented:
@phunkyfish I'm not able to find the mythtv pvr client, or any other pvr clients in the repository to install. 

The 10:08 screenshot is from my windows store kodi, and the 11:10 is from this build. I've browsed and searched, but haven't been able to find it.

![image](https://user-images.githubusercontent.com/28712901/76633556-4be45900-651b-11ea-9757-58359f610276.png)
![image](https://user-images.githubusercontent.com/28712901/76633655-70d8cc00-651b-11ea-9d39-1cdfe6b47cb6.png)

phunkyfish commented:
I think you need to rebase your change against master.

Do you know how to do that?

DerickLebron commented:
@phunkyfish I don't. Been clicking around and reading for a while here, but haven't figured it out.
I apologize, kind of out of my realm here. Still on tfs at work... I made these changes through the web browser a couple years ago so I don't have a local repository or IDE installed. If you want to just close this one and be done with it, I wouldn't complain since I obviously didn't follow the correct procedures.

phunkyfish commented:
Let me try one more build including the addons. But if you were not able to test this I'm not sure it's going to work first time!

phunkyfish commented:
Ok, try this one. And if it doesn't work we can call it a day ;)

https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20180925-63d4c7a5-patch-1-x64.exe

phunkyfish commented:
@MilhouseVH, could you add this one to a test build?

MilhouseVH commented:
Will be included in `#0311`.

(For any urgent test requests if would be best to ping me on Slack - my inbox is overflowing and I'm not paying it much attention these days...)

MilhouseVH commented:
Edit: `#0319`

metaron-uk commented:
I finally managed to roll my own (librelec 9.2 with this patch, rpi2 build) the day before the @MillhouseVH test build was released (I was testing it when the mail came in...).
Using the advancedsettings.xml below, I can confirm the ability to tweak when the commercial break cut starts and ends so that I get a suitable portion of the add break. I did try with values of 20 for wait and wind and that worked as well, but there was just too much add break!
I can also confirm that without the advancedsettings file, my commercial break behaviour is unchanged from stock libreelec 9.2. 
In my case, using silence detect on E4 star trek discovery recordings via mythtv in the UK, this means I get a bit of the add break after the cut, but I loose 1 or so seconds of the programme when it re starts, which I had been living with.

I’m happy with the patch and am keeping this build on my living room pi3 :)

I might try using a -ve value of < commbreakautowait > to see if I can trim the add duration back even further.

<advancedsettings>
<edl>
<!-- if true, commercial breaks will be merged according to the remaining options. -->
<mergeshortcommbreaks>true</mergeshortcommbreaks>
<!-- Commercial breaks won't be merged if the total length of the commercial break would be greater than this (seconds) -->
<maxcommbreaklength>250</maxcommbreaklength>
<!-- After merging, commercial breaks shorter than this will be removed (seconds) -->
<mincommbreaklength>90</mincommbreaklength>
<!-- Commercial breaks that are further apart than this won't be merged (seconds) -->
<maxcommbreakgap>120</maxcommbreakgap>
<!-- How long to wait before automatically skipping when the start of a commercial break reached (seconds) -->
<commbreakautowait>2</commbreakautowait>
<!-- How long to rewind after automatically skipping to the end of the commercial break (seconds) -->
<commbreakautowind>8</commbreakautowind>
</edl>
 <video>
  <timeseekforwardbig>300</timeseekforwardbig>  <!-- Time forward in sec doing a long seek.  Default 600 (10 minutes). -->
  <timeseekbackwardbig>-300</timeseekbackwardbig>  <!-- Time forward in secs when doing a long seek.  Default -600 (10 minutes). -->
 </video>
 <pictureextensions>
  <add>.mpo</add>
 </pictureextensions>
</advancedsettings>

> On 8 Mar 2020, at 19:54, phunkyfish <notifications@github.com> wrote:
> Cool, will post windows build shortly.
> 
> @MilhouseVH could you add this PR to the Rpi testbuilds so @metaron-uk can test also?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

metaron-uk commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -914,8 +914,8 @@ void CAdvancedSettings::ParseSettingsFile(const std::string &file)
     XMLUtils::GetInt(pElement, "maxstartgap", m_iEdlMaxStartGap, 0, 10 * 60);               // Between 0 and 10 minutes
-    XMLUtils::GetInt(pElement, "commbreakautowait", m_iEdlCommBreakAutowait, 0, 10);        // Between 0 and 10 seconds
-    XMLUtils::GetInt(pElement, "commbreakautowind", m_iEdlCommBreakAutowind, 0, 10);        // Between 0 and 10 seconds
+    XMLUtils::GetInt(pElement, "commbreakautowait", m_iEdlCommBreakAutowait, 0, 60);        // Between 0 and 60 seconds
```
changing to
```suggestion
    XMLUtils::GetInt(pElement, "commbreakautowait", m_iEdlCommBreakAutowait, -60, 60);        // Between -60 and 60 seconds
```
allows me to enter a value of -2 here in advancedsettings.xml
This allows me to trim off more of the commercial break so I only have to look at the pre and post commercial roll, rather than any of the commercials themselves.
Suggest allowing +/- 60 seconds for autowait and autowind will allow individual users to tweak the edl behaviour to their liking based on how well their particular back end’s commflagging method aligns with the add break start and end, and for personal taste regarding how much add they want to see to make sure they didn’t miss any programme.

metaron-uk commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -914,8 +914,8 @@ void CAdvancedSettings::ParseSettingsFile(const std::string &file)
-    XMLUtils::GetInt(pElement, "commbreakautowait", m_iEdlCommBreakAutowait, 0, 10);        // Between 0 and 10 seconds
-    XMLUtils::GetInt(pElement, "commbreakautowind", m_iEdlCommBreakAutowind, 0, 10);        // Between 0 and 10 seconds
+    XMLUtils::GetInt(pElement, "commbreakautowait", m_iEdlCommBreakAutowait, 0, 60);        // Between 0 and 60 seconds
+    XMLUtils::GetInt(pElement, "commbreakautowind", m_iEdlCommBreakAutowind, 0, 60);        // Between 0 and 60 seconds
```
```suggestion
    XMLUtils::GetInt(pElement, "commbreakautowind", m_iEdlCommBreakAutowind, -60, 60);        // Between -60 and 60 seconds
```

phunkyfish commented:
@DerickLebron can you update your PR with the changes?
Can help you with some directions of you don't know how. I also tried the changes and they work well.

DerickLebron commented:
@phunkyfish I think I did it. Sorry It took so long.

phunkyfish commented:
Ok, all you need to do now is squash your commits into a single commit.

If you have git on your PC and can run a few commands on the command line I can help you. If not I can try and do this for you.

DerickLebron commented:
@phunkyfish It turns out my brother uses git so he walked me through a squashing and rebasing with master.

phunkyfish commented:
Jenkins build this please

phunkyfish commented:
> @phunkyfish It turns out my brother uses git so he walked me through a squashing and rebasing with master.

Nice, well done. Once Jenkins is happy we should be good.

phunkyfish commented:
Jenkins build this please

DerickLebron commented:
@phunkyfish the android arm64 build failed. It doesn't look like the error was code related. Is this something that just happens periodically with the builds? Thanks for all your help with this.

phunkyfish commented:
Yes, we just have to build again until it’s all green.

phunkyfish commented:
Jenkins build this please

phunkyfish commented:
@DaVukovic does this require a Wiki update somewhere?

phunkyfish commented:
@DerickLebron well done, and thanks for the contribution ;)

DaVukovic commented:
I don't think so. advancedsettings was already documented. Only the values have changed for that setting, which I will mention and that wiki.

Maybe something needs to be added here: https://kodi.wiki/view/Edit_decision_list#MythTV
But I don't know much about MythTV and its backend to do that. Sorry

phunkyfish commented:
Ya, EDL's are definitely back in v19 alright. Clearly!
